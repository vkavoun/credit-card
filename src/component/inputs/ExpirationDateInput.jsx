import React from 'react';
import PropTypes from 'prop-types';

const yearOffset = 20;

export default function ExpirationDateInput({
  expiryMonth,
  expiryYear,
  handleExpiryMonthInput,
  handleExpiryYearInput,
}) {
  const currentDate = new Date(Date.now());
  const minYear = currentDate.getFullYear();
  const maxYear = minYear + yearOffset;

  const monthItems = [];
  const yearItems = [];

  for (let x = 1; x <= 12; x++) {
    monthItems.push(
      <option key={`month${x}`} value={x}>
        {x}
      </option>,
    );
  }

  for (let x = minYear; x <= maxYear; x++) {
    const value = parseInt(x.toString().substring(2), 10);
    yearItems.push(
      <option key={`year${value}`} value={value}>
        {value}
      </option>,
    );
  }

  return (
    <div>
      <label>Expiration Date*</label>
      <select
        id="card-expiry-month-manual-entry__keyed-entry-input"
        name="card-expiry-month-manual-entry__keyed-entry-input"
        data-testid="expiry-month"
        value={expiryMonth}
        onChange={e => handleExpiryMonthInput(e)}
      >
        {monthItems}
      </select>
      <select
        id="card-expiry-year-manual-entry__keyed-entry-input"
        name="card-expiry-year-manual-entry__keyed-entry-input"
        data-testid="expiry-year"
        value={expiryYear}
        onChange={e => handleExpiryYearInput(e)}
      >
        {yearItems}
      </select>
    </div>
  );
}

ExpirationDateInput.propTypes = {
  expiryMonth: PropTypes.number.isRequired,
  expiryYear: PropTypes.number.isRequired,
  handleExpiryMonthInput: PropTypes.func.isRequired,
  handleExpiryYearInput: PropTypes.func.isRequired,
};
