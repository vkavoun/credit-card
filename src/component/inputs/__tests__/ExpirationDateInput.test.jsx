import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import { render, fireEvent } from '@testing-library/react';
import ExpirationDateInput from '../ExpirationDateInput';

describe('ExpirationDateInput', () => {
  const currentDate = new Date(Date.now());
  const expiryMonth = currentDate.getFullYear();
  const expiryYear = currentDate.getMonth();

  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <ExpirationDateInput
        expiryMonth={expiryMonth}
        expiryYear={expiryYear}
        handleExpiryMonthInput={jest.fn()}
        handleExpiryYearInput={jest.fn()}
      />,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders correctly', () => {
    const tree = renderer
      .create(
        <ExpirationDateInput
          expiryMonth={expiryMonth}
          expiryYear={expiryYear}
          handleExpiryMonthInput={jest.fn()}
          handleExpiryYearInput={jest.fn()}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('ensure there are 12 items for months', () => {
    const { getByTestId } = render(
      <ExpirationDateInput
        expiryMonth={expiryMonth}
        expiryYear={expiryYear}
        handleExpiryMonthInput={jest.fn()}
        handleExpiryYearInput={jest.fn()}
      />,
    );
    expect(getByTestId('expiry-month')).toHaveLength(12);
  });

  it('ensure there are 21 items for years', () => {
    const { getByTestId } = render(
      <ExpirationDateInput
        expiryMonth={expiryMonth}
        expiryYear={expiryYear}
        handleExpiryMonthInput={jest.fn()}
        handleExpiryYearInput={jest.fn()}
      />,
    );
    expect(getByTestId('expiry-year')).toHaveLength(21);
  });
});
