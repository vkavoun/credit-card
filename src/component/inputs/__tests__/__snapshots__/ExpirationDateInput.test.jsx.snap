// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpirationDateInput renders correctly 1`] = `
<div>
  <label>
    Expiration Date*
  </label>
  <select
    data-testid="expiry-month"
    id="card-expiry-month-manual-entry__keyed-entry-input"
    name="card-expiry-month-manual-entry__keyed-entry-input"
    onChange={[Function]}
    value={2020}
  >
    <option
      value={1}
    >
      1
    </option>
    <option
      value={2}
    >
      2
    </option>
    <option
      value={3}
    >
      3
    </option>
    <option
      value={4}
    >
      4
    </option>
    <option
      value={5}
    >
      5
    </option>
    <option
      value={6}
    >
      6
    </option>
    <option
      value={7}
    >
      7
    </option>
    <option
      value={8}
    >
      8
    </option>
    <option
      value={9}
    >
      9
    </option>
    <option
      value={10}
    >
      10
    </option>
    <option
      value={11}
    >
      11
    </option>
    <option
      value={12}
    >
      12
    </option>
  </select>
  <select
    data-testid="expiry-year"
    id="card-expiry-year-manual-entry__keyed-entry-input"
    name="card-expiry-year-manual-entry__keyed-entry-input"
    onChange={[Function]}
    value={2}
  >
    <option
      value={20}
    >
      20
    </option>
    <option
      value={21}
    >
      21
    </option>
    <option
      value={22}
    >
      22
    </option>
    <option
      value={23}
    >
      23
    </option>
    <option
      value={24}
    >
      24
    </option>
    <option
      value={25}
    >
      25
    </option>
    <option
      value={26}
    >
      26
    </option>
    <option
      value={27}
    >
      27
    </option>
    <option
      value={28}
    >
      28
    </option>
    <option
      value={29}
    >
      29
    </option>
    <option
      value={30}
    >
      30
    </option>
    <option
      value={31}
    >
      31
    </option>
    <option
      value={32}
    >
      32
    </option>
    <option
      value={33}
    >
      33
    </option>
    <option
      value={34}
    >
      34
    </option>
    <option
      value={35}
    >
      35
    </option>
    <option
      value={36}
    >
      36
    </option>
    <option
      value={37}
    >
      37
    </option>
    <option
      value={38}
    >
      38
    </option>
    <option
      value={39}
    >
      39
    </option>
    <option
      value={40}
    >
      40
    </option>
  </select>
</div>
`;
